haystack/telemetry.py

```diff
@@ -0,0 +1,343 @@
+    """
+    if "google.colab" in sys.modules:
+        execution_env = "colab"
+    elif "KUBERNETES_SERVICE_HOST" in os.environ:
+        execution_env = "kubernetes"
+    elif HAYSTACK_DOCKER_CONTAINER in os.environ:
+        execution_env = os.environ.get(HAYSTACK_DOCKER_CONTAINER)
+    # check if pytest is imported
+    elif "pytest" in sys.modules:
+        execution_env = "test"
+    else:
+        try:
+            shell = get_ipython().__class__.__name__
+            execution_env = shell
+        except Exception:
+            execution_env = "script"
+    return execution_env
+
+
+def _read_telemetry_config():
+    """
+    Loads the config from the file specified in CONFIG_PATH
+    """
+    global user_id
+    try:
+        if not Path(CONFIG_PATH).is_file():
+            return
+        with open(CONFIG_PATH, "r", encoding="utf-8") as stream:
+            config = yaml.safe_load(stream)
+            if "user_id" in config and user_id is None:
+                user_id = config["user_id"]
+    except Exception:
+        pass
+
+
+def _write_telemetry_config():
+    """
+    Writes a config file storing the randomly generated user id and whether to write events to a log file.
+    This method logs an info to inform the user about telemetry when it is used for the first time.
+    """
+    global user_id
+    try:
+        # show a log message if telemetry config is written for the first time
+        if not Path(CONFIG_PATH).is_file():
+            logger.info(
+                f'Haystack sends anonymous usage data to understand the actual usage and steer dev efforts towards features that are most meaningful to users. You can opt out at anytime by setting {HAYSTACK_TELEMETRY_ENABLED}="False" as an environment variable or by calling disable_telemetry(). More information at https://haystack.deepset.ai/overview/faq'
+            )
+            Path(CONFIG_PATH).parents[0].mkdir(parents=True, exist_ok=True)
+        user_id = _get_or_create_user_id()
+        config = {"user_id": user_id}
+
+        with open(CONFIG_PATH, "w") as outfile:
+            yaml.dump(config, outfile, default_flow_style=False)
+    except Exception:
+        send_custom_event(event="config saving failed")
+
+
+def _delete_telemetry_config():
+    """
+    Deletes the telemetry config file if any exists.
+    """
+    try:
+        Path(CONFIG_PATH).unlink(missing_ok=True)
+    except Exception:
+        pass
+
+
+def _write_event_to_telemetry_log_file(distinct_id: str, event: str, properties: Dict[str, Any]):
+    try:
+        with open(LOG_PATH, "a") as file_object:
+            file_object.write(f"{event}, {properties}, {distinct_id}\n")
+    except Exception:
+        pass
+
+
+def _delete_telemetry_log_file():
+    """
+    Deletes the telemetry log file if any exists.
+    """
+    try:
+        Path(LOG_PATH).unlink(missing_ok=True)
+    except Exception:
+        pass
+
+
+class NonPrivateParameters:
```

Comment:
What you describe is definitely an alternative. For now, I'd say we don't want to allow editing the list of non private parameters arbitrarily. Removing from that list would be okay but adding to the list is not.  Removing from that list would make analysis more difficult. I'd say we keep it as is. Maybe more functionality is added over time to the `NonPrivateParameters` class. In that case, it would makes sense to keep to separate the code a bit from the other functionality of telemetry.

Reply:
No problem :+1: 

Just a note: right now that list is editable too, so in practice the change has no impact on how "private" it is. This is just a style topic though and I highly doubt this has any impact whatsoever, so no need to discuss further :grin: